File: LeetCode/69. Sqrt(x)/69.cpp
#include <cmath>

class Solution {
public:
    int mySqrt(int x) {
        if (x < 2) return x;

        long long x0 = x, x1;
        while (true) {
            x1 = (x0 + x / x0) / 2;
            if (x0 - x1 < 1)
                break;
            x0 = x1;
        }
        return x0;
    }
};

File: LeetCode/69. Sqrt(x)/69.py
class Solution:
    def mySqrt(self, x):
        if x < 2:
            return x
        
        left, right = 1, x // 2
        while left <= right:
            mid = (left + right) // 2
            squared = mid * mid
            if squared == x:
                return mid
            elif squared < x:
                left = mid + 1
            else:
                right = mid - 1
        
        return right  # right will be the largest mid where mid*mid <= x
